### DISCUSSION TITLE
Mark one or many routes to a web api controller endpoints as deprecated/obsolete

### DISCUSSION THREAD

**Discussion by SiDiamond** [dotnet/csharplang#8258] [Language Ideas]
In web api - where there are multiple route attributes on a controller method, can we please have the ability to mark one or many of the routes as obsolete/deprecated  Currently to achieve this I have to duplicate the method and make one of the methods obsolete/deprecated.  I must have one method for obsolete/deprecated routes, and one method for non obsolete/deprecated routes. I have hundreds of endpoints like this in my app. The code duplication is horrible, 

Current code with duplication:

`[ApiController]
[Route("api/[controller]")]
public class AuditController : ControllerBase
{

    [Obsolete]
    [HttpGet("v1/auditHistory"),]
    public IActionResult GetAuditHistory()
    {
        return Ok("This route output is the same for both routes, but one route is obsolete");
    }

    [HttpGet("v2/auditHistory")]
    public IActionResult GetAuditHistory()
    {
        return Ok("This route output is the same for both routes, but one route is obsolete");
    }
}`

New suggestion:

`[ApiController]
[Route("api/[controller]")]
public class AuditController : ControllerBase
{
    [HttpGet("v1/auditHistory"), Obsolete]
    [HttpGet("v2/auditHistory")]
    public IActionResult GetAuditHistory()
    {
        return Ok("This route output is the same for both routes, but one route is obsolete");
    }
}`

Thank you :)

**Comment by colejohnson66** [comment]
This is (1) not a language issue, but an ASP one, and (2) is not possible because it would change how attributes work. Attributes apply to the method, so your proposal would mark the method as obsolete regardless of where the attribute is placed.

Also, why are you marking API route *methods* obsolete? That doesn't do anything to the client.

If you want one method that handles both routes, just put the `HttpGetAttribute` twice:

```csharp
[HttpGet("v1/auditHistory")]
[HttpGet("v2/auditHistory")]
public IActionResult GetAuditHistory() => Ok("");
```

If you want separate methods, you don't need to duplicate code. Just forward the method call! If you ever find yourself duplicating code, ask yourself if there's an easier solution.

```csharp
[HttpGet("v1/auditHistory")]
[Obsolete]
public IActionResult GetAuditHistoryOld() => GetAuditHistory();

[HttpGet("v2/auditHistory")]
public IActionResult GetAuditHistory() => Ok("");
```

**Reply by SiDiamond** [reply]
Hi. Thank you for responding... I don't want separate methods. I would like to have multiple routes to the same method, and mark one or more as obsolete. This stems from URL changes to endpoints over time. 

When using Swashbuckle to auto generate swagger for the api endpoint it will use the multiple route attributes to generate separate API endpoint definitions (these can then be imported to an API manager like Azure APIM), but you cant mark just one of the routes obsolete as all routs to that method become obsolete (as you mention - the whole method becomes obsolete). To present the old route as obsolete and add a new route you must duplicate the method. OK, one can use the method that you have shown to call one method from another (as we already do), but then there is a lot of duplication of other attributes, parameters, XML mark-up etc etc that is needed by both methods. It would be very convenient if there was a way of marking a single route as obsolete and not the entire method.

**Reply by colejohnson66** [reply]
Again. Your proposal will not work because it would *break* existing code's obsoletion markers. The *only* possible solution would be an extra property on `HttpGetAttribute`, but that's not a language concern - it's an ASP.NET request and should be asked there.

**Comment by HaloFour** [comment]
Combining attributes in this manner is already legal:

```cs
[Foo]
[Bar, Baz]
public void M() { }
```

It means the same thing as the following:

```cs
[Foo]
[Bar]
[Baz]
public void M() { }

// or 

[Foo, Bar, Baz]
public void M() { }
```

If you're looking for a way to denote that a particular application of an attribute is obsolete, I would make a request to the owner of that project to include an `Obsolete` property for that attribute:

```cs
[Foo(Obsolete = true)]
[Bar]
[Baz]
public void M() { }
```

**Reply by SiDiamond** [reply]
Thank you for responding... Yes, I'm aware of the combining of attributes. It would be great to have 

[HttpGet("v1/auditHistory"), ObsoleteRoute = true] 

or similar, only denoting that this particular route is obsolete, and not marking the method and all it's routes as obsolete (as it does now).
